src/main/java/org/opentripplanner/graph_builder/linking/StreetSplitter.java

```diff
@@ -1,11 +1,834 @@
+     * @param traverseModeSet The traverse modes.
+     * @param options The routing options.
+     * @param destructiveSplitting If true splitting is permanent (Used when linking transit stops etc.) when
+     *                             false the original edge is kept in the graph after creating split edges.
+     * @param createSemiPermanentEdges If true, semi-permanent edges and a semi-permanent splitter vertex is created.
+     *                                 These vertices and edges are intended to be kept in the graph and linked to
+     *                                 during subsequent requests, but might be removed at a later time. For example a
+     *                                 floating bike rental may need to be removed from the graph once the data is
+     *                                 updated to show that the floating bike rental no longer exists or if its
+     *                                 location has significantly changed.
+     */
+    public boolean linkToGraph(
+        Vertex vertex,
+        TraverseModeSet traverseModeSet,
+        RoutingRequest options,
+        final boolean destructiveSplitting,
+        boolean createSemiPermanentEdges
+    ) {
+        // find nearby street edges
+        // TODO: we used to use an expanding-envelope search, which is more efficient in
+        // dense areas. but first let's see how inefficient this is. I suspect it's not too
+        // bad and the gains in simplicity are considerable.
+        final double radiusDeg = SphericalDistanceLibrary.metersToDegrees(MAX_SEARCH_RADIUS_METERS);
+
+        Envelope env = new Envelope(vertex.getCoordinate());
+
+        // Perform a simple local equirectangular projection, so distances are expressed in degrees latitude.
+        final double xscale = Math.cos(vertex.getLat() * Math.PI / 180);
+
+        // Expand more in the longitude direction than the latitude direction to account for converging meridians.
+        env.expandBy(radiusDeg / xscale, radiusDeg);
+
+        final double DUPLICATE_WAY_EPSILON_DEGREES = SphericalDistanceLibrary.metersToDegrees(DUPLICATE_WAY_EPSILON_METERS);
+
+        // We sort the list of candidate edges by distance to the stop
+        // This should remove any issues with things coming out of the spatial index in different orders
+        // Then we link to everything that is within DUPLICATE_WAY_EPSILON_METERS of of the best distance
+        // so that we capture back edges and duplicate ways.
+        List<StreetEdge> candidateEdges = idx.query(env).stream()
+            .filter(
+                streetEdge -> streetEdge instanceof StreetEdge &&
+                    // do not find SemiPermanentPartialStreetEdges if creating new SemiPermanentPartialStreetEdges
+                    // use only permanent StreetEdges instead
+                    (createSemiPermanentEdges
+                        ? !(streetEdge instanceof SemiPermanentPartialStreetEdge)
+                        : true)
```

Comment:
Ternary could be simplified to `(!createSemiPermanentEdges || !(streetEdge instanceof SemiPermanentPartialStreetEdge))`, but I'm not sure that's more readable than what's here, so let's just leave as is.

Reply:
Will refactor, and also fix the above comment.